File: src/main/scala/com/vc/medium/L286.scala
package com.vc.medium

object L286 {
  def wallsAndGates(rooms: Array[Array[Int]]): Unit = {
    import scala.collection.mutable
    val q = new mutable.Queue[(Int, Int)]()
    rooms.indices.foreach(i => {
      rooms(i).indices.foreach(j => {
        if(rooms(i)(j) == 0) q.enqueue((i, j))
      })
    })

    val n = rooms.length
    val m = if(n > 0) rooms(0).length else 0
    while(q.nonEmpty) {
      val p = q.dequeue()
      val row = p._1
      val col = p._2
      if(row > 0 && rooms(row - 1)(col) == Int.MaxValue) {
        rooms(row - 1)(col) = rooms(row)(col) + 1
        q.enqueue((row - 1, col))
      }
      if(row < n - 1 && rooms(row + 1)(col) == Int.MaxValue) {
        rooms(row + 1)(col) = rooms(row)(col) + 1
        q.enqueue((row + 1, col))
      }
      if(col > 0 && rooms(row)(col - 1) == Int.MaxValue) {
        rooms(row)(col - 1) = rooms(row)(col) + 1
        q.enqueue((row, col - 1))
      }
      if(col < m - 1 && rooms(row)(col + 1) == Int.MaxValue) {
        rooms(row)(col + 1) = rooms(row)(col) + 1
        q.enqueue((row, col + 1))
      }
    }
  }
}

File: src/main/java/com/vc/medium/L286.java
package com.vc.medium;

import java.util.*;

class L286 {
    private int[][] rooms;

    public void wallsAndGates(int[][] rooms) {
        this.rooms = rooms;

        int n = rooms.length;
        if(n == 0) return;

        int m = rooms[0].length;
        if(m == 0) return;

        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(rooms[i][j] == 0) {
                    bfs(i, j);
                }
            }
        }
    }

    private int[][] dirs = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    private void bfs(int row, int col) {
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{row, col});

        while(!q.isEmpty()) {
            int[] e = q.poll();
            //System.out.println(e[0]+" "+e[1]);
            for(int[] dir: dirs) {
                int rowNew = e[0] + dir[0];
                int colNew = e[1] + dir[1];

                if(rowNew >= 0 && rowNew < rooms.length && colNew >= 0
                        && colNew < rooms[0].length) {
                    if(rooms[rowNew][colNew] == -1) continue;
                    if(rooms[rowNew][colNew] > rooms[e[0]][e[1]] + 1) {
                        rooms[rowNew][colNew] = rooms[e[0]][e[1]] + 1;
                        q.offer(new int[]{rowNew, colNew});
                    }
                }
            }
        }
    }
}
